File: src/advent_of_code_2015/Day11/main.cs
using System.Runtime.CompilerServices;
using System.Text;

class Program
{
    static void Main()
    {   
        string password = "cqjxxyzz";
        // while (!IsValidPassword(password))
        // {
        password = IncrementPassword(password);
        while (!IsValidPassword(password))
        {
            password = IncrementPassword(password);
        }
        //}

        Console.WriteLine(password);
    }

    private static bool IsValidPassword(string password)
    {
        //rule 1: can't contain i, o, l
        if (password.Contains('i') || password.Contains('o')  || password.Contains('l') )
        {
            return false;
        }

        // Rule 2: Must have one increasing straight of 3 letters
        bool hasStraight = false;
        for (int i = 0; i < password.Length - 2; i++)
        {
            if (password[i + 1] == password[i] + 1 && password[i + 2] == password[i] + 2)
            {
                hasStraight = true;
                break;
            }
        }

        if (!hasStraight) return false;

        // Rule 3: Must have two non-overlapping pairs
        int pairCount = 0;
        for (int i = 0; i < password.Length - 1; i++)
        {
            if (password[i] == password[i + 1])
            {
                pairCount++;
                i++; // skip next to avoid overlapping
            }
        }

        return pairCount >= 2;
    }

    private static string IncrementPassword(string password)
    {
        char[] chars = password.ToCharArray();
        int i = chars.Length - 1;
        
        while (i >= 0)
        {
            if (chars[i] == 'z')
            {
                chars[i] = 'a';
                i--;
            }
            else
            {
                chars[i]++;
                break;
            }
        }

        return new string(chars);
    }
}

File: src/advent_of_code_2015/Day11/main.py
def increment_password(input: str) -> str:
    char_list = input
    for i in range(len(input) - 1, -1, -1): 
        if char_list[i] == 'z':
            char_list[i] = 'a'
            i -= 1
        else:
            char_list[i] += 1
            break
    
    return "".join(char_list)

def is_valid_password(input: str) -> bool: 
    char_list = input
    for i in range(len(input)):
        if input[i] == 'i' or input[i] == 'o' or input[i] == 'l':
            return False
    
    has_straight = False
    for i in range(len(input) - 2):
        if input[i + 1] == input[i] + 1 and input[i + 2] == input[i] + 2:
            has_straight = True
            break
    
    if has_straight is False:
        return False
    
    pairCount = 0
    for i in range(len(input) - 1):
        if input[i] == input[i + 1]:
            pairCount += 1
            i += 1
            
    return  pairCount >= 2




password = "cqjxjnds"

while is_valid_password is not True: 
    password = increment_password(password)

print(password)